<a href="https://colab.research.google.com/github/RV05/AWS-ML-DL_projects/blob/main/Commit_1_ALL_Global_Terrorism_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> The Global Terrorism Database (GTD) is an open-source database including information on terrorist attacks around the world from 1970 through 2017. The GTD includes systematic data on domestic as well as international terrorist incidents that have occurred during this time period and now includes more than 180,000 attacks. The database is maintained by researchers at the National Consortium for the Study of Terrorism and Responses to Terrorism (START), headquartered at the University of Maryland.</b>

#<b>Problem Statement</b>
## <b> Explore and analyze the data to discover key findings pertaining to terrorist activities. </b>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import numpy as np
import pandas as pd




In [ ]:
df=pd.read_csv("/content/drive/MyDrive/Global Terrorism Data (1).csv",encoding='ISO-8859-1')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (4,6,31,33,61,62,63,76,79,90,92,94,96,114,115,121) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [ ]:
df.rename(columns={'iyear':'Year','imonth':'Month','iday':'Day','country_txt':'Country','provstate':'state','region_txt':'Region','attacktype1_txt':'AttackType','target1':'Target','nkill':'Killed','nwound':'Wounded','summary':'Summary','gname':'Group','targtype1_txt':'Target_type','weaptype1_txt':'Weapon_type','motive':'Motive','success':'success'},inplace=True)

In [ ]:
df.shape

(181691, 135)

In [ ]:
df.describe()

eventid           Year  ...       INT_MISC        INT_ANY
count  1.816910e+05  181691.000000  ...  181691.000000  181691.000000
mean   2.002705e+11    2002.638997  ...       0.090010      -3.945952
std    1.325957e+09      13.259430  ...       0.568457       4.691325
min    1.970000e+11    1970.000000  ...      -9.000000      -9.000000
25%    1.991021e+11    1991.000000  ...       0.000000      -9.000000
50%    2.009022e+11    2009.000000  ...       0.000000       0.000000
75%    2.014081e+11    2014.000000  ...       0.000000       0.000000
max    2.017123e+11    2017.000000  ...       1.000000       1.000000

[8 rows x 77 columns]

In [ ]:
df.head()

eventid  Year  Month  Day  ... INT_IDEO  INT_MISC INT_ANY  related
0  197000000001  1970      7    2  ...        0         0       0      NaN
1  197000000002  1970      0    0  ...        1         1       1      NaN
2  197001000001  1970      1    0  ...       -9         1       1      NaN
3  197001000002  1970      1    0  ...       -9         1       1      NaN
4  197001000003  1970      1    0  ...       -9         1       1      NaN

[5 rows x 135 columns]

In [ ]:
100-(((df.loc[:,'approxdate'].count())/181691)*100)

94.91499303762983

about 94.9% of values are missing in this column. Before dropping these values, let's see if there is a specific pattern, meaning are these values missing in between a particular date, or missing for a particular location etc

This gives me an intuition that maybe other columns have just as many null 
values. I need to get a sense of the number of null values that are there in the dataset in a metric (likely %absoulte) that gives me an idea of what I am dealing with.


Before that let's drop the columns that won't contribute to our problem statement.


While trying to drop the columns, problem faced is I don't know the names of all the columns. so printing them first.

In [ ]:
print(df.columns.values)


['eventid' 'Year' 'Month' 'Day' 'approxdate' 'extended' 'resolution'
 'country' 'Country' 'region' 'Region' 'state' 'city' 'latitude'
 'longitude' 'specificity' 'vicinity' 'location' 'Summary' 'crit1' 'crit2'
 'crit3' 'doubtterr' 'alternative' 'alternative_txt' 'multiple' 'success'
 'suicide' 'attacktype1' 'AttackType' 'attacktype2' 'attacktype2_txt'
 'attacktype3' 'attacktype3_txt' 'targtype1' 'Target_type' 'targsubtype1'
 'targsubtype1_txt' 'corp1' 'Target' 'natlty1' 'natlty1_txt' 'targtype2'
 'targtype2_txt' 'targsubtype2' 'targsubtype2_txt' 'corp2' 'target2'
 'natlty2' 'natlty2_txt' 'targtype3' 'targtype3_txt' 'targsubtype3'
 'targsubtype3_txt' 'corp3' 'target3' 'natlty3' 'natlty3_txt' 'Group'
 'gsubname' 'gname2' 'gsubname2' 'gname3' 'gsubname3' 'Motive'
 'guncertain1' 'guncertain2' 'guncertain3' 'individual' 'nperps'
 'nperpcap' 'claimed' 'claimmode' 'claimmode_txt' 'claim2' 'claimmode2'
 'claimmode2_txt' 'claim3' 'claimmode3' 'claimmode3_txt' 'compclaim'
 'weaptype1' 'Weapon_typ

In [ ]:
100-(((df.loc[:,'approxdate'].count())/181691)*100)

94.91499303762983

% of values missing in each column

In [ ]:
dict_of_null_values={}
for columns in df.columns:
  dict_of_null_values[columns]=100-(((df.loc[:,columns].count())/181691)*100)
print(dict_of_null_values)

{'eventid': 0.0, 'Year': 0.0, 'Month': 0.0, 'Day': 0.0, 'approxdate': 94.91499303762983, 'extended': 0.0, 'resolution': 98.7781453126462, 'country': 0.0, 'Country': 0.0, 'region': 0.0, 'Region': 0.0, 'state': 0.23171208260177423, 'city': 0.2388670875277228, 'latitude': 2.507554034046805, 'longitude': 2.5081044190411177, 'specificity': 0.00330230996581804, 'vicinity': 0.0, 'location': 69.45638474112641, 'Summary': 36.39640928829716, 'crit1': 0.0, 'crit2': 0.0, 'crit3': 0.0, 'doubtterr': 0.0005503849942982697, 'alternative': 84.03278093026071, 'alternative_txt': 84.03278093026071, 'multiple': 0.0005503849942982697, 'success': 0.0, 'suicide': 0.0, 'attacktype1': 0.0, 'AttackType': 0.0, 'attacktype2': 96.5248691459676, 'attacktype2_txt': 96.5248691459676, 'attacktype3': 99.7644352224381, 'attacktype3_txt': 99.7644352224381, 'targtype1': 0.0, 'Target_type': 0.0, 'targsubtype1': 5.709143545910365, 'targsubtype1_txt': 5.709143545910365, 'corp1': 23.41888150761457, 'Target': 0.3500448563770391

In [ ]:
null_values_df=pd.DataFrame.from_dict(dict_of_null_values, orient='index')

In [ ]:
null_values_df.rename(columns={0:'% of null values'},inplace=True)

In [ ]:
null_values_df[null_values_df['% of null values']>80].describe()

% of null values
count         68.000000
mean          96.560474
std            4.148616
min           84.032781
25%           94.018003
50%           98.941885
75%           99.668255
max           99.988992

We have 68 rows which have more than 80% values as null values

Let's drop the columns which have more than 80%
But before we do that let's see if we can reduce the error by dropping columns before a specific date

In [ ]:
null_values_df['1997_null_values']=df[df['Year']>1997].isnull().sum()*100/len(df[df['Year']>1997])

In [ ]:
null_values_df['2002_null_values']=df[df['Year']>2002].isnull().sum()*100/len(df[df['Year']>2002])

In [ ]:
null_values_df['2007_null_values']=df[df['Year']>2007].isnull().sum()*100/len(df[df['Year']>2007])

In [ ]:
null_values_df['2012_null_values']=df[df['Year']>2012].isnull().sum()*100/len(df[df['Year']>2012])

In [ ]:
null_values_df['2017_null_values']=df[df['Year']>2017].isnull().sum()*100/len(df[df['Year']>2017])

In [ ]:
null_values_df.head()

% of null values  ...  2017_null_values
eventid             0.000000  ...               NaN
Year                0.000000  ...               NaN
Month               0.000000  ...               NaN
Day                 0.000000  ...               NaN
approxdate         94.914993  ...               NaN

[5 rows x 6 columns]

In [ ]:
null_values_df.describe()

% of null values  1997_null_values  ...  2012_null_values  2017_null_values
count        135.000000        135.000000  ...        135.000000               0.0
mean          56.481719         53.047855  ...         53.167977               NaN
std           44.118525         45.705381  ...         45.856431               NaN
min            0.000000          0.000000  ...          0.000000               NaN
25%            0.220429          0.003941  ...          0.000000               NaN
50%           84.032781         83.597527  ...         81.297247               NaN
75%           98.941885         98.737126  ...         98.530508               NaN
max           99.988992         99.995621  ...        100.000000               NaN

[8 rows x 6 columns]

if we look at the 25% row, it can be inferenced that droppping columns from 2012 section, and about 25%, will lead to least loss of features

In [ ]:
null_values_df.head()

% of null values  ...  2017_null_values
eventid             0.000000  ...               NaN
Year                0.000000  ...               NaN
Month               0.000000  ...               NaN
Day                 0.000000  ...               NaN
approxdate         94.914993  ...               NaN

[5 rows x 6 columns]

In [ ]:
df['Year'].value_counts()

2014    16903
2015    14965
2016    13587
2013    12036
2017    10900
2012     8522
2011     5076
1992     5071
2010     4826
2008     4805
2009     4721
1991     4683
1989     4324
1990     3887
1988     3721
1984     3495
1994     3456
2007     3242
1997     3197
1987     3183
1995     3081
1996     3058
1985     2915
1983     2870
1986     2860
2006     2758
1979     2662
1980     2662
1981     2586
1982     2544
2005     2017
2001     1906
2000     1814
1978     1526
1999     1395
2002     1333
1977     1319
2003     1278
2004     1166
1998      934
1976      923
1975      740
1970      651
1974      581
1972      568
1973      473
1971      471
Name: Year, dtype: int64

In [ ]:
clean_df=df[df['Year']>2012]

In [ ]:
clean_df.describe()

eventid          Year  ...      INT_MISC       INT_ANY
count  6.839100e+04  68391.000000  ...  68391.000000  68391.000000
mean   2.014984e+11   2014.918293  ...      0.062523     -4.354813
std    1.333243e+08      1.334403  ...      0.242105      4.718135
min    2.013010e+11   2013.000000  ...      0.000000     -9.000000
25%    2.014042e+11   2014.000000  ...      0.000000     -9.000000
50%    2.015043e+11   2015.000000  ...      0.000000     -9.000000
75%    2.016071e+11   2016.000000  ...      0.000000      0.000000
max    2.017123e+11   2017.000000  ...      1.000000      1.000000

[8 rows x 77 columns]

In [ ]:
clean_df.head()

eventid  Year  ...  INT_ANY                     related
113299  201301010001  2013  ...       -9                         NaN
113300  201301010002  2013  ...       -9  201301010002, 201301010003
113301  201301010003  2013  ...       -9  201301010002, 201301010003
113302  201301010004  2013  ...       -9                         NaN
113303  201301010005  2013  ...        0                         NaN

[5 rows x 135 columns]

In [ ]:
dict_of_null_values_clean={}
for columns in clean_df.columns:
  dict_of_null_values_clean[columns]=100-(((clean_df.loc[:,columns].count())/68391)*100)
print(dict_of_null_values_clean)

{'eventid': 0.0, 'Year': 0.0, 'Month': 0.0, 'Day': 0.0, 'approxdate': 87.47203579418344, 'extended': 0.0, 'resolution': 100.0, 'country': 0.0, 'Country': 0.0, 'region': 0.0, 'Region': 0.0, 'state': 0.002924361392572905, 'city': 0.0, 'latitude': 0.6243511573160134, 'longitude': 0.6243511573160134, 'specificity': 0.0, 'vicinity': 0.0, 'location': 55.963503969820586, 'Summary': 0.0, 'crit1': 0.0, 'crit2': 0.0, 'crit3': 0.0, 'doubtterr': 0.001462180696293558, 'alternative': 81.29724671374889, 'alternative_txt': 81.29724671374889, 'multiple': 0.001462180696293558, 'success': 0.0, 'suicide': 0.0, 'attacktype1': 0.0, 'AttackType': 0.0, 'attacktype2': 95.41606351712944, 'attacktype2_txt': 95.41606351712944, 'attacktype3': 99.67247152403094, 'attacktype3_txt': 99.67247152403094, 'targtype1': 0.0, 'Target_type': 0.0, 'targsubtype1': 7.06233276308285, 'targsubtype1_txt': 7.06233276308285, 'corp1': 0.0, 'Target': 0.0, 'natlty1': 1.242853591846881, 'natlty1_txt': 1.242853591846881, 'targtype2': 91.

In [ ]:
null_df_clean=pd.DataFrame.from_dict(dict_of_null_values_clean, orient='index')

In [ ]:
null_df_clean[null_df_clean>80].describe()

0
count   69.000000
mean    95.362705
std      5.502744
min     81.297247
25%     91.573453
50%     98.467635
75%     99.483850
max    100.000000

In [ ]:
null_values_df[null_values_df['% of null values']>80].describe()

% of null values  1997_null_values  ...  2012_null_values  2017_null_values
count         68.000000         68.000000  ...         68.000000               0.0
mean          96.560474         95.491178  ...         95.293348               NaN
std            4.148616          5.378218  ...          5.857860               NaN
min           84.032781         76.152526  ...         76.135749               NaN
25%           94.018003         91.818249  ...         91.529222               NaN
50%           98.941885         98.737126  ...         98.530508               NaN
75%           99.668255         99.511096  ...         99.505052               NaN
max           99.988992         99.995621  ...        100.000000               NaN

[8 rows x 6 columns]